### DISCUSSION TITLE
Add a new recipe by importing an image of it

### DISCUSSION THREAD

**Discussion by strenkml** [mealie-recipes/mealie#660] [Ideas]
This would remove the need for having to manually type in recipes that are in a recipe book or in a picture that someone shares with you.  Google's Tesseract would probably be the best way to accomplish this.  Getting the text in the image to a string would probably be the easy part.  The hard part would be parsing the data to find the title, description. ingredients, method, etc.  Since there are tons of different formats that a recipe could use this would be difficult.  

I have a couple of idea of how this could be done but I am not sure what the best route would be.  Does anyone have any experience with OCR?  I am willing to try implementing this but I do not have any experience with OCR so it would take some trial and error.

**Comment by tweek11** [comment]
This would be so useful. Was any thought or progress made on it. I’ve got allot family recipes typed out on paper (just not the time to retype them)

**Reply by hay-kot** [reply]
Yes, checkout https://demo.mealie.io/recipe/create/ocr

The UI is there but could use some UX work.

**Reply by tweek11** [reply]
Thank you for the timely reply and my apology - Currently running Version .56 and didn't realize this was a feature implemented in the Beta version. Guess now is as good as any time to switch over to the beta version.

**Comment by Plevuus** [comment]
I tried this feature today. It worked well for an english recipe, and reasonable for a Dutch recipe. Making it accept .jpg files would make it much better usable!

**Comment by crispybegs** [comment]
> I tried this feature today. It worked well for an english recipe, and reasonable for a Dutch recipe. Making it accept .jpg files would make it much better usable!

apologies for bumping an old thread, but does this ocr feature appear in the latest version? I can't seem to locate it.

**Reply by aureateflux** [reply]
It's still there as an option when you manually create a recipe, though there was some drama over it the other month, during which hay-kot mentioned it might be axed (because it wasn't being maintained by the contributor).

**Reply by cardonator** [reply]
Hopefully not because it has great potential. I wonder if there is any open source library that could help make it more accurate for handwriting. The current solution is atrocious at hand written recipes. Also, it would be nice if you could import multiple scans. I have scanned the front and back of my Mom's old recipe box handwritten on 3x5 index cards and being able to add multiple images to a single recipe would be amazing.

**Comment by rtrnms** [comment]
in v1.2.0 this option seems to be no longer available. can anyone confirm this? and any plans on bringing this back? it would be so wonderful!

**Reply by boc-the-git** [reply]
Correct, we removed OCR a couple months back. We had no one interested in supporting the feature and it added a ton of overhead to the application in general.

IIRC, a community member added an iOS guide for how to setup a shortcut to do this via AI. (don't quote me on the specifics, I don't use it)

**Comment by rtrnms** [comment]
Thanks a lot for the prompt reply!
You're right, there's a community article on the iOS shortcut for this
feature. Bad thing only that I'm an Android user.
I'm thinking about some work around e.g., converting pictures by OCR to
pdf, then to HTML and from there use the html page as URL to import to
mealie...

boc-the-git ***@***.***> schrieb am Sa., 17. Feb. 2024, 18:41:

**Reply by boc-the-git** [reply]
> Bad thing only that I'm an Android user.

You and me both!

I wanted to eventually use the OCR feature myself, but at the end of the day, it didn't make sense for us to maintain it as part of the app.
Please do share anything you sort out in this space, as I (and others) may want to use the same/similar!

**Reply by indytechcook** [reply]
Is it possible to share the Prompt used for the iOS shortcut?  We could probably create something for Andriod that would do something similar.

**Comment by probert94** [comment]
Is it possible to import a whole recipe with plain text? This way you could use an external picture to text software and just paste the text into mealie.

**Reply by bverkron** [reply]
You can paste in a list of ingredients and parse them but have to manually copy / paste steps individually or enter them by hand AFAIK. I don’t see a way to paste in an entire recipe with steps and ingredients and have it parsed.

**Comment by KevinKTM** [comment]
Following. I just set up Mealie for my wife and the first recipe she tried to add was from a cookbook. OCR would be a fantastic addition to this app (we are loving it so far!)

**Comment by Ge0rg3** [comment]
Hey, wondering if maybe some GPT-4 call could be made to extract the recipes? Would be much more reliable than just OCR, but would require users to enter an OpenAI key.

**Reply by paulsgale** [reply]
I just did this in Chat GPT-4 with a photo of an old family recipe that was typed and faxed many years ago (also was covered in chocolate icing smears!) I asked it to create a recipe and it worked perfectly. Cut and pasted into Mealie. Would be amazing to have the power of GPT/large language models - imagine the possibilities.

**Reply by probert94** [reply]
@paulsgale How did you paste the recipe into mealie? Did you paste every part (description, ingredients, steps) on it's own?
It would be nice to have some kind of text import for websites that don't support an URL import

**Reply by pgale** [reply]
I just separated the ingredients and method and pasted those in separately. It worked but I agree 100% that there needs to be a better text import parsing method. 

I've also found quite a few simple websites don't work with the URL import so that part could do with some improvements. When it works, it's great though and I really like the system but seems quite immature in those areas maybe.

**Reply by pgale** [reply]
I also gave chat GPT-4 a URL to a fairly complex article recipe (the Guardian newspaper) and asked it to pull out the ingredients and method. First off, it searched for other recipes and combined them and I had to ask it to specifically just use the URL I provided. It did that perfectly too. 

Not sure if GPT is the right tool but it certainly works well. Now that there's an API, that would seem to have interesting possibilities.

**Comment by indytechcook** [comment]
I asked Google Gemini how to use Google Gemini to import images to Mealie.  As the conversation continued, I added the possible use of Tandoor's OCR tooling.

Here is the conversation:  https://g.co/gemini/share/449e5d162111

I'm thinking through a solution.

**Comment by michael-genson** [comment]
The newest nightly has the ability to import via image using OpenAI: https://github.com/mealie-recipes/mealie/pull/3974; it can also optionally translate the recipe for you (with varying success depending on which model you use)

**Reply by pgale** [reply]
Oh that's amazing. Thanks. Looking forward to trying that out :)

**Reply by michael-genson** [reply]
Let me know how it works for you if you do! It's still very new so there's likely plenty of room for improvement

**Reply by Plevuus** [reply]
I've been busy in last months with other things than recipies and Mealie, but I just upgraded and tested this feature. It's great!! I really like it. Thanks a lot!!
